Avoid infinite integration by parts in manualintegrate

normalhuman opened the pull request:
#### References to other Issues or PRs

Fixes #9858 

#### Brief description of what is fixed or changed

With an integral like `exp(x) * cos(exp(x))`, manualintegrate attempts integration by parts with `u = cos(exp(x))`, which "succeeds", but leads to an infinite chain of different integrals such as `exp(n*x) * cos(exp(x))` and `exp(n*x) * sin(exp(x))`. While repeated integration with the same `dv` term can be useful (like `x**9 * exp(x)`), the same is not true for repeated use of `u` term, i.e., differentiating a factor and getting the same factor back. This is now prevented by keeping track of u-terms used by integration by parts.
 
#### Release Notes
 
<!-- BEGIN RELEASE NOTES -->
* integrals
  * prevented creation of infinite chains of integration by parts 
<!-- END RELEASE NOTES -->

normalhuman committed:
Avoid infinite integration by parts in manualintegrate

With an integral like `exp(x) * cos(exp(x))`, manualintegrate attempts
integration by parts with `u = cos(exp(x))`, which "succeeds", but
leads to an infinite chain of different integrals of the form `exp(n*x)
* cos(exp(x))` and `exp(n*x) * sin(exp(x))`. While repeated integration
with the same `dv` term can be useful (like `x**9 * exp(x)`), the same
is not true for repeated use of `u` term, i.e., differentiating a
factor and getting the same factor back. This is now prevented by
keeping track of u-terms used by integration by parts.

jksuom commented on sympy/integrals/manualintegrate.py:
```diff
@@ -496,6 +496,12 @@ def parts_rule(integral):
 
+        # Set a limit on the number of times u can be used
+        cachekey = u.xreplace({symbol: _cache_dummy})
+        if _parts_u_cache[cachekey] > 5:
```
I tried this with bound 2. It seems to work as well.

asmeurer replied:
This only catches infinite recursion from the same u term. I wonder if it can happen even with this logic, from slightly different terms each time. My guess is not currently because it only handles LIATE terms but would that continue to be the case if we extend it (e.g., https://github.com/sympy/sympy/issues/15431).

normalhuman replied:
Differentiating functions other than trigonometric / exponential is unlikely to return the integral to its original form. This check is meant for these types only. 

With the more targeted check, I reduced the bound to 2 as suggested by @jksuom, and added a test to make sure that `manualintegrate` is still able to do a bunch of similar integrations by parts.  

I don't know how to prevent manualintegrate from creating an infinite sequence of integrands (which is also possible with other methods, like substitution) without imposing an arbitrary limit on the recursion depth (which would hurt with integrands like `x**n * exp(x)`).

normalhuman committed:
More specific restriction on repeated u-partss

Limit the restriction on the repeated use of the same u-part of
integration by parts to the function prone to reproducing themselves
when differentiated. Smaller limit (2 instead of 5) to avoid going
through futile motions. A test to show that multiple integration by
parts is still possible.

normalhuman committed:
remove extra blank line

jksuom commented on sympy/integrals/manualintegrate.py:
```diff
@@ -541,6 +548,7 @@ def make_second_step(steps, integrand):
         if (constant != 1) and rule:
             rule = ConstantTimesRule(constant, integrand, rule,
                                      constant * integrand, symbol)
+        cachekey = u.xreplace({symbol: _cache_dummy})
```
Where is this used?

normalhuman replied:
I've no idea how this line got there. :(

normalhuman committed:
remove a stray line of code

normalhuman committed:
empty commit to restart tests

jksuom commented:
I think this should work for most integrands. If it should turn out that the bound is too low, we could think of giving it as a keyword parameter.
